crates/holochain_integrity_types/src/chain.rs

```diff
@@ -131,47 +141,89 @@ impl<H: Eq + Ord + std::hash::Hash> ChainFilter<H> {
+                StopConditions::UntilTimestamp(timestamp.max(old_ts))
+            }
+            StopConditions::UntilHash(u) => StopConditions::Multiple(u32::MAX, u, timestamp),
+            StopConditions::Multiple(n, u, _) => StopConditions::Multiple(n, u, timestamp),
+        };
+        self
+    }
+
     /// Take all actions until this action hash is found.
-    /// Note that all actions specified as `until` hashes must be
+    /// Note that all actions specified as `until_hash` hashes must be
     /// found so this filter can produce deterministic results.
     /// It is invalid to specify an until hash that is on a different
     /// fork then the starting position.
-    pub fn until(mut self, action_hash: H) -> Self {
-        self.filters = match self.filters {
-            ChainFilters::ToGenesis => ChainFilters::Until(Some(action_hash).into_iter().collect()),
-            ChainFilters::Take(n) => ChainFilters::Both(n, Some(action_hash).into_iter().collect()),
-            ChainFilters::Until(mut u) => {
+    pub fn until_hash(mut self, action_hash: H) -> Self {
+        self.stop_conditions = match self.stop_conditions {
+            StopConditions::ToGenesis => {
+                StopConditions::UntilHash(Some(action_hash).into_iter().collect())
+            }
+            StopConditions::Take(n) => {
+                StopConditions::Multiple(n, Some(action_hash).into_iter().collect(), Timestamp(0))
+            }
+            StopConditions::UntilTimestamp(t) => {
+                StopConditions::Multiple(u32::MAX, Some(action_hash).into_iter().collect(), t)
+            }
+            StopConditions::UntilHash(mut u) => {
                 u.insert(action_hash);
-                ChainFilters::Until(u)
+                StopConditions::UntilHash(u)
             }
-            ChainFilters::Both(n, mut u) => {
+            StopConditions::Multiple(n, mut u, t) => {
                 u.insert(action_hash);
-                ChainFilters::Both(n, u)
+                StopConditions::Multiple(n, u, t)
             }
         };
         self
     }
 
     /// Get the until hashes if there are any.
-    pub fn get_until(&self) -> Option<&HashSet<H>> {
-        match &self.filters {
-            ChainFilters::Until(u) => Some(u),
-            ChainFilters::Both(_, u) => Some(u),
+    pub fn get_until_hash(&self) -> Option<&HashSet<H>> {
+        match &self.stop_conditions {
+            StopConditions::UntilHash(u) => Some(u),
+            StopConditions::Multiple(_, u, _) => Some(u),
             _ => None,
         }
     }
 
     /// Get the take number if there is one.
     pub fn get_take(&self) -> Option<u32> {
-        match &self.filters {
-            ChainFilters::Take(s) => Some(*s),
-            ChainFilters::Both(s, _) => Some(*s),
+        match &self.stop_conditions {
+            StopConditions::Take(n) => Some(*n),
+            StopConditions::Multiple(n, _, _) => {
+                if *n != u32::MAX {
```

Comment:
This is a logic change to an existing piece of code that appears unrelated to the main change on this PR. This needs calling out so it doesn't get missed. I also don't understand why this change is being made, so the call out for it needs to include a justification.

Reply:
Looking much better as logic with `Option`, I'm finding that much easier to follow